OptimizelySDK/Bucketing/DecisionService.cs

```diff
@@ -0,0 +1,237 @@
+        public DecisionService(Bucketer bucketer, IErrorHandler errorHandler, ProjectConfig projectConfig, UserProfileService userProfileService, ILogger logger)
+        {
+            Bucketer = bucketer;
+            ErrorHandler = errorHandler;
+            ProjectConfig = projectConfig;
+            UserProfileService = userProfileService;
+            Logger = logger;
+        }
+
+        /// <summary>
+        /// Get a Variation of an Experiment for a user to be allocated into.
+        /// </summary>
+        /// <param name = "experiment" > The Experiment the user will be bucketed into.</param>
+        /// <param name = "userId" > The userId of the user.
+        /// <param name = "filteredAttributes" > The user's attributes. This should be filtered to just attributes in the Datafile.</param>
+        /// <returns>The Variation the user is allocated into.</returns>
+        public Variation GetVariation(Experiment experiment, string userId, UserAttributes filteredAttributes)
+        {
+            if (!ExperimentUtils.IsExperimentActive(experiment, Logger)) return null;
+
+            var variation = GetWhitelistedVariation(experiment, userId);
+
+            if (variation != null)   return variation;
+
+            if (UserProfileService != null)
+            {
+                try
+                {
+                    Dictionary<string, object> userProfileMap = UserProfileService.Lookup(userId);
+                    if (UserProfileUtils.IsValidUserProfileMap(userProfileMap))
+                    {
+                        var userProfile = UserProfileUtils.ConvertMapToUserProfile(userProfileMap);
+                        variation = GetStoredVariation(experiment, userProfile);
+                        if (variation != null) return variation;
+                    }
+                    else if (userProfileMap == null)
+                    {
+                        Logger.Log(LogLevel.INFO, "We were unable to get a user profile map from the UserProfileService.");
+                    }
+                    else
+                    {
+                        Logger.Log(LogLevel.ERROR, "The UserProfileService returned an invalid map.");
+                    }
+                }
+                catch (Exception exception)
+                {
+                    Logger.Log(LogLevel.ERROR, exception.Message);
+                    ErrorHandler.HandleError(new Exceptions.OptimizelyRuntimeException(exception.Message));
+                }
+            }
+
+            if (ExperimentUtils.IsUserInExperiment(ProjectConfig, experiment, filteredAttributes))
+            {
+                
+                variation = Bucketer.Bucket(ProjectConfig, experiment, userId);
+
+                if (variation != null && variation.Key != null)
+                {
+                    if (UserProfileService != null)
+                        SaveVariation(experiment, variation, new UserProfile(userId, new Dictionary<string, Decision>()));
```

Comment:
We shouldn't create a new user profile every time we save. There could be an existing one with decisions for other experiments. We need to preserve those. I suggest keeping a copy of the user profile returned on line 83 and only creating a new user profile if that one is null